### ISSUE TITLE
lines-changed-only option does not work correctly

### ISSUE THREAD

**Issue by turol** [cpp-linter/cpp-linter-action#74]
`lines-changed-only` option does not work correctly and sometimes complains about lines that were not changed by the PR. For example this PR:
https://github.com/chocolate-doom/chocolate-doom/pull/1480

Elicited comment "File opl/opl.c (lines 42, 56): Code does not conform to Custom style guidelines."
Line 42 was not changed. Line 56 wasn't either, but line 55 was.

https://github.com/chocolate-doom/chocolate-doom/runs/7307673916

Correct(ish) option appears to have been passed to clang-format: `--lines=50:58` though the range should have been 53:55. Line 42 should definitely not have raised issues.

**Comment by 2bndy5** [comment]
While I agree with you, I just want to point out that this might be clang-format's fault. Lets examine the complete object from the diff (starting at line 42):
```diff
 static opl_driver_t *drivers[] =
 {
 #if (defined(__i386__) || defined(__x86_64__)) && defined(HAVE_IOPERM)
     &opl_linux_driver,
 #endif
 #if defined(HAVE_LIBI386) || defined(HAVE_LIBAMD64)
     &opl_openbsd_driver,
 #endif
 #ifdef _WIN32
     &opl_win32_driver,
 #endif
+#ifndef DISABLE_SDL2MIXER
     &opl_sdl_driver,
+#endif // DISABLE_SDL2MIXER
     NULL
 };
```
So, the correct line numbers were passed to clang format, but it probably flagged line 42 as well because it was the start of the object definition. 

Passing line ranges to clang-format doesn't necessarily mean it will only examine those lines. It has to examine the whole file and then filter out the warnings to only those _related_ to the specified range.

---
I think we can have an extra check in the action src to ensure [the lines mentioned](https://github.com/cpp-linter/cpp-linter-action/blob/d24e6225d827632732599cb812817ac086781cfd/cpp_linter/clang_format_xml.py#L113) are only in the diff. But this may be difficult if clang-format is nesting the warnings for diff'd lines within warnings for non-diff'd lines - its all [parsed from clang-format's XML output](https://github.com/cpp-linter/cpp-linter-action/blob/d24e6225d827632732599cb812817ac086781cfd/cpp_linter/clang_format_xml.py#L127-L129).

**Comment by turol** [comment]
Could you add a new option to restrict it more to only lines changed? The point was that we would want it to complain only about changed lines. The head maintainer didn't want to fix all the formatting at once.

**Comment by 2bndy5** [comment]
TL:DR This isn't a simple request.

clang-format's `--lines` arg only accepts ranges.
>```
>  --lines=<string>           - <start line>:<end line> - format a range of
>                               lines (both 1-based).
>                               Multiple ranges can be formatted by specifying
>                               several -lines arguments.
>``` 

I think the intention of such an option should focus on additions only. Getting the line number for a added line in the diff is rather difficult because the diff doesn't actually have line number at the start of the lines in the diff. Meaning, line numbers have to be literally counted from the beginning of the diff, and subtracted lines in the diff _will_ negatively offset the actual line number for added lines. 

They literally write entire libraries just for parsing diffs, but none were intuitive enough to satisfy the needs for such an option (I already looked into it). This is why I settled for focusing on just what's shown in the diff.

**Comment by 2bndy5** [comment]
BTW, that's a neat project. I haven't played doom since my win95 machine. Now, I just use the retroarch port when I'm feeling nostalgic.

**Comment by 2bndy5** [comment]
It seems I already tried to filter the clang-format warnings.
https://github.com/cpp-linter/cpp-linter-action/blob/d24e6225d827632732599cb812817ac086781cfd/cpp_linter/thread_comments.py#L131-L136

Still digging into this...

**Comment by shenxianpeng** [comment]
> The head maintainer didn't want to fix all the formatting at once.

But if possible, perform a full format then ignore this commit with git blame is a common practice, refer to [llvm-project](https://github.com/llvm/llvm-project/blob/main/.git-blame-ignore-revs).

**Comment by 2bndy5** [comment]
ok, I think I've got the `strict` additions feature implemented, but I need to test it on an actual CI event. I haven't yet added the extra filtering of warnings per only lines affected by commit.

So, far, the unit test looks promising...

**Comment by 2bndy5** [comment]
I've extended the linefiltering to the `thread_comments` feature (as well as the initially requested `file_annotations` feature). This will be solved when backlogged updates branch is merged into main.

BTW, @turol I used the commit from the chocolate-doom repo (referenced in this issue's OP) as case for the new unit tests. I hope you don't mind.

**Comment by turol** [comment]
@2bndy5 I don't mind but the chocolate-doom code is GPLed so you might want to check with a lawyer if that causes a problem.

**Comment by 2bndy5** [comment]
The code isn't actually being distributed with this repo, but the changed files (all 9 of them) from that commit are downloaded every time we run the unit tests. I'll check if this is suitable.

**Comment by 2bndy5** [comment]
Looks like the license is focused on distribution, and anything related to copying the code seems to be subject to modifications.
1. We are not distributing the code as part of this repo. the tests do download some files from chocolate-doom, but they are explicitly gitignored. There will never be a copy of the files/code from chocolate-doom project in this repo.
2. We are not modifying the code, just analyzing it to make sure this issue is resolved.

In order to run the unit tests without hosting the code, I've copied the event_payload for the commit referenced in the OP of this thread. This event_payload.json does not contain any information that can be used maliciously. I have removed all info that isn't necessary to running the test (which doesn't require anything about users or user access). See https://github.com/cpp-linter/cpp-linter-action/commit/6a1f91d2e62230fb9cacf1e741d680b8d3bc72b6

If I've misinterpreted the [GPL for chocolate-doom](https://github.com/chocolate-doom/chocolate-doom/blob/master/COPYING.md), then we may have to find a different sample to perform unit tests on.

**Comment by 2bndy5** [comment]
After further consideration, I think the newer "strict" value should be the default (when `lines-changed-only` is `'true'`) and the old behavior should be used only if `lines-changed-only` is set to `'diff'`. Given that the user intuition (in OP) assumed this would be the case, it might be seen as an improved default.
